docs/proposals/unified-backend-architecture/tech-spec.md

```diff
@@ -0,0 +1,221 @@
+
+**TODO**: To clarify details of receive address generation. Will it be generated on the client side on on the node
+side? If latter - we need additional API endpoint, something like `/api/v1/address/receive/new`, to obtain new
+receive address.
+
+### Send a Payment
+
+There are 3 steps to make a new payment:
+
+1. create transaction,
+2. sign transaction,
+3. publish transaction.
+
+When user creates a new payment (by defining recipient's address, amount and fee), it produces a `POST`-request
+to `/api/v1/transactions/unsigned`, and response will contain a new transaction without a witness (signature).
+
+After that user signs new transaction using its private keys (hosted on the client side).
+
+Then user send a payment. It produces a `POST`-request to `/api/v1/transactions/signed`, node receives transaction
+with a signature, forms transaction witness and publish this transaction to the Cardano network as usually.
+
+## API Extension
+
+API v1 must be extended to support both the mobile clients and the hardware wallets.
+
+### Create Unsigned Transaction
+
+When user wants to send money to other user, he creates a new payment. It produces a `POST`-request
+to `/api/v1/transactions/unsigned` endpoint, and response will contain a new transaction without a
+witness (signature).
+
+### Send Signed Transaction
+
+After user signs new transaction using its private keys (hosted on the client side) he sends it with
+a signature. It produces a `POST`-request to `/api/v1/transactions/signed`, node receives transaction
+with a signature, forms transaction witness and publish this transaction as usually.
+
+## Wallet State
+
+An analysis of the computational complexity of rebuilding the state of a wallet (e.g. its balance and
+payments history) on the server in a way as much stateless as possible.
+
+**TODO**: To think about it.
+
+## Authentication Scheme
+
+### HTTPS
+
+Client must be sure that server is not a malicious one, so we use HTTPS-connection with a standard check
+of TLS certs on the client side.
+
+### Session
+
+After successful TLS checking client and server establish a new **session**.
+
+**TODO**: To think what is a session? What data is associated with the session? Does it have any timeout?
+
+## Crypto Operations
+
+A description of how the main operations on a crypto-wallet will be implemented, according to the final
+index derivation mechanism chosen (sequential vs random).
+
+**TODO**: To think about it.
+
+## Security and Possible Attacks
+
+### MITM
+
+To prevent MITM attacks we have to include a fee in transaction explicitly.
+
+Currently it is impossible, because transaction contains inputs, outputs and attributes, and we have to
+know _previous_ transaction to calculate a fee. So we must change transaction's format for explicit
+including a fee in it.
+
+**TODO**: To think about it: should the fee be a part of transaction's attributes, or it should be a
+separate field?
+
+### Other Attacks
+
+**TODO**: To think about it.
```

Comment:
@vincenthz What do you think about other possible attacks?

Reply:
This should eventually also be reviewed by the security internal auditors.